homeassistant/components/aemet/weather_update_coordinator.py

```diff
@@ -0,0 +1,655 @@
+    get_forecast_day_value,
+    get_forecast_hour_value,
+    get_forecast_interval_value,
+)
+import async_timeout
+
+from homeassistant.components.weather import (
+    ATTR_FORECAST_CONDITION,
+    ATTR_FORECAST_PRECIPITATION,
+    ATTR_FORECAST_PRECIPITATION_PROBABILITY,
+    ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME,
+    ATTR_FORECAST_WIND_BEARING,
+    ATTR_FORECAST_WIND_SPEED,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_API_CONDITION,
+    ATTR_API_FORECAST,
+    ATTR_API_HUMIDITY,
+    ATTR_API_PRESSURE,
+    ATTR_API_RAIN,
+    ATTR_API_RAIN_PROB,
+    ATTR_API_SNOW,
+    ATTR_API_SNOW_PROB,
+    ATTR_API_STATION_ID,
+    ATTR_API_STATION_NAME,
+    ATTR_API_STATION_TIMESTAMP,
+    ATTR_API_STORM_PROB,
+    ATTR_API_TEMPERATURE,
+    ATTR_API_TEMPERATURE_FEELING,
+    ATTR_API_TOWN_ID,
+    ATTR_API_TOWN_NAME,
+    ATTR_API_TOWN_TIMESTAMP,
+    ATTR_API_WIND_BEARING,
+    ATTR_API_WIND_MAX_SPEED,
+    ATTR_API_WIND_SPEED,
+    CONDITIONS_MAP,
+    DOMAIN,
+    FORECAST_MODE_DAILY,
+    FORECAST_MODE_HOURLY,
+    WIND_BEARING_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+WEATHER_UPDATE_INTERVAL = timedelta(minutes=10)
+
+
+def format_condition(condition: str) -> str:
+    """Return condition from dict CONDITIONS_MAP."""
+    for key, value in CONDITIONS_MAP.items():
+        if condition in value:
+            return key
+    return condition
+
+
+def format_float(value) -> float:
+    """Try converting string to float."""
+    try:
+        return float(value)
+    except ValueError:
+        return None
+
+
+def format_int(value) -> int:
+    """Try converting string to int."""
+    try:
+        return int(value)
+    except ValueError:
+        return None
+
+
+class WeatherUpdateCoordinator(DataUpdateCoordinator):
+    """Weather data update coordinator."""
+
+    def __init__(self, aemet, latitude, longitude, forecast_mode, hass):
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=WEATHER_UPDATE_INTERVAL
+        )
+
+        self._aemet = aemet
+        self._station = None
+        self._town = None
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecast_mode = forecast_mode
+        self._weather_data = None
+        self._data = {
+            "daily": None,
+            "hourly": None,
+            "station": None,
+        }
+
+    async def _async_update_data(self):
+        data = {}
+        try:
+            with async_timeout.timeout(120):
+                weather_response = await self._get_aemet_weather()
+                data = self._convert_weather_response(weather_response)
+                self._weather_data = data
+        except asyncio.TimeoutError:
+            data = self._weather_data
+            _LOGGER.error("AEMET OpenData API timeout")
```

Comment:
Just let the timeout error propagate here. The standard handling is to log an error and set last_update_success = False.